Creating a user with a non-existent secret causes a panic

Create a CassandraDatacenter like this:

```yaml
apiVersion: cassandra.datastax.com/v1beta1
kind: CassandraDatacenter
metadata:
  name: dc1
spec:
  clusterName: cluster1
  serverType: cassandra
  serverVersion: 3.11.7
  managementApiAuth:
    insecure: {}
  size: 3
  storageConfig:
    cassandraDataVolumeClaimSpec:
      storageClassName: server-storage
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 5Gi
  config:
    cassandra-yaml:
      authenticator: org.apache.cassandra.auth.PasswordAuthenticator
      authorizer: org.apache.cassandra.auth.CassandraAuthorizer
      role_manager: org.apache.cassandra.auth.CassandraRoleManager
    jvm-options:
      initial_heap_size: 800M
      max_heap_size: 800M
  users:
  - secretName: does-not-exist
    superuser: true
```

The `validateCassandraUserSecrets` function in `secrets.go` does not check that the secret actually exists. This in turn causes a panic like this:

```
E0121 03:45:47.364093       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 1190 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1570740, 0x25047f0)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1570740, 0x25047f0)
        /usr/local/go/src/runtime/panic.go:969 +0x166
github.com/datastax/cass-operator/operator/pkg/reconciliation.validateCassandraUserSecretContent(0xc0001ac800, 0x0, 0x0, 0x0, 0x1)
        /cass-operator/operator/pkg/reconciliation/secrets.go:231 +0x3a
github.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconciliationContext).validateCassandraUserSecrets(0xc0000c28c0, 0xc0006ebaf0, 0x0, 0x0)
        /cass-operator/operator/pkg/reconciliation/secrets.go:286 +0x1df
github.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconciliationContext).isValid(0xc0000c28c0, 0xc0001ac800, 0xc000319410, 0xc0001b6a10)
        /cass-operator/operator/pkg/reconciliation/handler.go:301 +0x9f
github.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconcileCassandraDatacenter).Reconcile(0xc0007368c0, 0xc000789f99, 0x7, 0xc0007895ec, 0x3, 0x0, 0x0, 0x0, 0x0)
        /cass-operator/operator/pkg/reconciliation/handler.go:245 +0x773
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001aaa80, 0x15cc3c0, 0xc0006b6900, 0xc000793600)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:256 +0x161
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0001aaa80, 0xc000793700)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:232 +0xae
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0001aaa80)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00048b2f0)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:152 +0x5f
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00048b2f0, 0x3b9aca00, 0x0, 0xc000453401, 0xc00059db60)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc00048b2f0, 0x3b9aca00, 0xc00059db60)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:193 +0x305
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x136f4ba]

goroutine 1190 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/runtime/runtime.go:55 +0x105
panic(0x1570740, 0x25047f0)
        /usr/local/go/src/runtime/panic.go:969 +0x166
github.com/datastax/cass-operator/operator/pkg/reconciliation.validateCassandraUserSecretContent(0xc0001ac800, 0x0, 0x0, 0x0, 0x1)
        /cass-operator/operator/pkg/reconciliation/secrets.go:231 +0x3a
github.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconciliationContext).validateCassandraUserSecrets(0xc0000c28c0, 0xc0006ebaf0, 0x0, 0x0)
        /cass-operator/operator/pkg/reconciliation/secrets.go:286 +0x1df
github.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconciliationContext).isValid(0xc0000c28c0, 0xc0001ac800, 0xc000319410, 0xc0001b6a10)
        /cass-operator/operator/pkg/reconciliation/handler.go:301 +0x9f
github.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconcileCassandraDatacenter).Reconcile(0xc0007368c0, 0xc000789f99, 0x7, 0xc0007895ec, 0x3, 0x0, 0x0, 0x0, 0x0)
        /cass-operator/operator/pkg/reconciliation/handler.go:245 +0x773
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001aaa80, 0x15cc3c0, 0xc0006b6900, 0xc000793600)
```

--- a/operator/pkg/reconciliation/secrets.go
+++ b/operator/pkg/reconciliation/secrets.go
@@ -149,7 +149,7 @@ func (rc *ReconciliationContext) createInternodeCACredential() (*corev1.Secret, 
 	}
 	if keypem, certpem, err := utils.GetNewCAandKey(fmt.Sprintf("%s-ca-keystore", rc.Datacenter.Name), rc.Datacenter.Namespace); err == nil {
 		secret.Data = map[string][]byte{
-			"key": []byte(keypem),
+			"key":  []byte(keypem),
 			"cert": []byte(certpem),
 		}
 		return secret, nil
@@ -158,11 +158,11 @@ func (rc *ReconciliationContext) createInternodeCACredential() (*corev1.Secret, 
 	}
 }
 
-func (rc *ReconciliationContext) createCABootstrappingSecret(jksBlob []byte) (error) {
+func (rc *ReconciliationContext) createCABootstrappingSecret(jksBlob []byte) error {
 	_, err := rc.retrieveSecret(types.NamespacedName{
-			Name:      fmt.Sprintf("%s-keystore", rc.Datacenter.Name),
-			Namespace: rc.Datacenter.Namespace,
-		})
+		Name:      fmt.Sprintf("%s-keystore", rc.Datacenter.Name),
+		Namespace: rc.Datacenter.Namespace,
+	})
 
 	if err == nil { // This secret already exists, nothing to do
 		return nil
@@ -227,19 +227,23 @@ func (rc *ReconciliationContext) retrieveInternodeCredentialSecretOrCreateDefaul
 func validateCassandraUserSecretContent(dc *api.CassandraDatacenter, secret *corev1.Secret) []error {
 	var errs []error
 
-	namespacedName := types.NamespacedName{
-		Name:      secret.ObjectMeta.Name,
-		Namespace: secret.ObjectMeta.Namespace,
-	}
-	errorPrefix := fmt.Sprintf("Validation failed for user secret: %s", namespacedName.String())
-
-	for _, key := range []string{"username", "password"} {
-		value, ok := secret.Data[key]
-		if !ok {
-			errs = append(errs, fmt.Errorf("%s Missing key: %s", errorPrefix, key))
-		} else if !utf8.Valid(value) {
-			errs = append(errs, fmt.Errorf("%s Key did not have valid utf8 value: %s", errorPrefix, key))
+	if secret != nil {
+		namespacedName := types.NamespacedName{
+			Name:      secret.ObjectMeta.Name,
+			Namespace: secret.ObjectMeta.Namespace,
 		}
+		errorPrefix := fmt.Sprintf("Validation failed for user secret: %s", namespacedName.String())
+
+		for _, key := range []string{"username", "password"} {
+			value, ok := secret.Data[key]
+			if !ok {
+				errs = append(errs, fmt.Errorf("%s Missing key: %s", errorPrefix, key))
+			} else if !utf8.Valid(value) {
+				errs = append(errs, fmt.Errorf("%s Key did not have valid utf8 value: %s", errorPrefix, key))
+			}
+		}
+	} else {
+		errs = append(errs, fmt.Errorf("Validation failed due to a missing secret"))
 	}
 
 	return errs
@@ -273,14 +277,16 @@ func (rc *ReconciliationContext) validateCassandraUserSecrets() []error {
 	for _, user := range users {
 		secretName := user.SecretName
 		namespace := dc.ObjectMeta.Namespace
-
-		secret, err := rc.retrieveSecret(types.NamespacedName{
+		secretKey := types.NamespacedName{
 			Name:      secretName,
 			Namespace: namespace,
-		})
+		}
+
+		secret, err := rc.retrieveSecret(secretKey)
 
 		if err != nil {
 			errs = append(errs, fmt.Errorf("Validation of user secret failed due to an error: %w", err))
+			continue
 		}
 
 		errs = append(errs, validateCassandraUserSecretContent(dc, secret)...)

--- a/operator/pkg/reconciliation/secrets_test.go
+++ b/operator/pkg/reconciliation/secrets_test.go
@@ -113,6 +113,13 @@ func Test_validateCassandraUserSecretContent(t *testing.T) {
 			valid:   false,
 			message: "validation should fail when secret contains non-utf8 data",
 		},
+		{
+			superuserSecret: "my-fun-secret",
+			secretNil:       true,
+			data:            nil,
+			valid:           false,
+			message:         "validation should fail when secret does not exists",
+		},
 	}
 
 	for _, test := range tests {

